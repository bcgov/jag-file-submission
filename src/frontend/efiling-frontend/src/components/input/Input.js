/* eslint-disable react/jsx-one-expression-per-line */
import React from "react";
import PropTypes from "prop-types";
import "./Input.scss";

export const Input = ({
  input: {
    label,
    id,
    value,
    placeholder,
    isRequired,
    errorMsg,
    styling,
    note,
    isReadOnly,
    isControlled,
  },
  onChange,
}) => {
  let asterisk = "";

  if (isRequired) {
    asterisk = (
      <span id="asterisk" className="bcgov-mandatory">
        *
      </span>
    );
  }

  let labelPart = null;
  const labelPartExists = label !== false;

  if (label) {
    labelPart = (
      <label htmlFor={id} className="bcgov-label">
        <div className="bcgov-input-label">{label}</div>
        {asterisk}&nbsp;
        <span className="bcgov-note">{note}</span>
      </label>
    );
  }

  return (
    <div>
      {labelPartExists && labelPart}
      {isControlled && (
        <input
          className={styling}
          type="text"
          id={id}
          value={value}
          data-testid="input-test"
          placeholder={placeholder}
          readOnly={isReadOnly}
          onChange={(event) => onChange(event.target.value)}
        />
      )}
      {!isControlled && (
        <input
          className={styling}
          type="text"
          id={id}
          defaultValue={value}
          data-testid="input-test"
          placeholder={placeholder}
          readOnly={isReadOnly}
          onChange={(event) => onChange(event.target.value)}
        />
      )}
      <br />
      <span className="bcgov-error" data-testid="email-error">{errorMsg}</span>
    </div>
  );
};

Input.propTypes = {
  input: PropTypes.shape({
    label: PropTypes.string,
    id: PropTypes.string.isRequired,
    defaultValue: PropTypes.string,
    value: PropTypes.string,
    placeholder: PropTypes.string,
    isRequired: PropTypes.bool.isRequired,
    errorMsg: PropTypes.string,
    styling: PropTypes.string.isRequired,
    note: PropTypes.string,
    isReadOnly: PropTypes.bool,
    isControlled: PropTypes.bool,
  }).isRequired,
  onChange: PropTypes.func,
};

Input.defaultProps = {
  onChange: () => {},
};
