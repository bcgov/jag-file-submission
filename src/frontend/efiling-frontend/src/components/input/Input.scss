.bcgov-label {
  display: flex;
  flex-wrap: wrap;
  align-items: baseline;
}

.bcgov-input-label {
  font-weight: 700;
  font-style: normal;
  font-size: 16px;
  color: #333333;
  text-align: left;
  line-height: 16px;
}

.bcgov-editable-white {
  padding: 1px 0px 1px 10px;
  box-sizing: border-box;
  top: 30px;
  width: 100%;
  height: 40px;
  background-color: white;
  font-weight: 400;
  font-style: normal;
  font-size: 16px;
  text-decoration: none;
  color: #494949;
  text-align: left;
  border-style: inset;
}

.bcgov-non-editable-grey {
  padding: 1px 0px 1px 10px;
  top: 30px;
  width: 100%;
  height: 40px;
  background-color: #f2f2f2;
  font-weight: 400;
  font-style: normal;
  font-size: 16px;
  text-decoration: none;
  color: #494949;
  text-align: left;
  border-style: solid;
  border-color: #b1b1b1;
  border-width: 2px;
}

.bcgov-note {
  font-size: 14px;
  font-weight: 400;
}

.bcgov-error {
  color: red;
}

.bcgov-mandatory {
  padding: 0px 0px 0px 3px;
  color: red;
  font-weight: 500;
  font-size: 16px;
}
