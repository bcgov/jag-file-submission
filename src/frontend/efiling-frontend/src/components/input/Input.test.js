import React from "react";
import { render, fireEvent, getByRole } from "@testing-library/react";
import { Input } from "./Input";
import testBasicSnapshot from "../../TestHelper";

describe("Input Component", () => {
  const input = {
    styling: "bcgov-editable-white",
    id: "textInputId",
    isReadOnly: false,
  };

  const onChange = jest.fn();

  test("when mandatory and with label matches the snapshot (editable white)", () => {
    const inputElement = (
      <Input
        input={{
          ...input,
          isRequired: true,
          placeholder: "Enter id",
          label: "Submission ID",
        }}
        onChange={onChange}
      />
    );

    testBasicSnapshot(inputElement);
  });

  test("when not mandatory and with label matches the snapshot (editable white)", () => {
    const inputElement = (
      <Input
        input={{
          ...input,
          isRequired: false,
          placeholder: "Enter id",
          label: "Submission ID",
        }}
        onChange={onChange}
      />
    );

    testBasicSnapshot(inputElement);
  });

  test("when mandatory and no label matches the snapshot (editable white)", () => {
    const inputElement = (
      <Input
        input={{
          ...input,
          isRequired: true,
          placeholder: "Enter id",
        }}
        onChange={onChange}
      />
    );

    testBasicSnapshot(inputElement);
  });

  test("when not mandatory and no label matches the snapshot (editable white)", () => {
    const inputElement = (
      <Input
        input={{
          ...input,
          isRequired: false,
          placeholder: "Enter id",
        }}
        onChange={onChange}
      />
    );

    testBasicSnapshot(inputElement);
  });

  test("matches the snapshot (non-editable-grey)", () => {
    const inputElement = (
      <Input
        input={{
          ...input,
          isRequired: false,
          isReadOnly: true,
          placeholder: "Enter id",
        }}
        onChange={onChange}
      />
    );

    testBasicSnapshot(inputElement);
  });

  test("on change updates the value in the input field", () => {
    const { container } = render(
      <Input
        input={{
          ...input,
          isRequired: false,
          placeholder: "Enter id",
        }}
        onChange={onChange}
      />
    );

    fireEvent.change(getByRole(container, "textbox"), {
      target: { value: "examplevalue" },
    });

    expect(getByRole(container, "textbox").value).toBe("examplevalue");
  });
});
