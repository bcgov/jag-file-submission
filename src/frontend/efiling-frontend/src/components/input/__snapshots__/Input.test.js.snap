// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component matches the snapshot (non-editable-grey) 1`] = `
<div>
  <input
    className="bcgov-editable-white"
    data-testid="input-test"
    id="textInputId"
    onChange={[Function]}
    placeholder="Enter id"
    readOnly={true}
    type="text"
  />
  <br />
  <span
    className="bcgov-error"
    data-testid="email-error"
  />
</div>
`;

exports[`Input Component when mandatory and no label matches the snapshot (editable white) 1`] = `
<div>
  <input
    className="bcgov-editable-white"
    data-testid="input-test"
    id="textInputId"
    onChange={[Function]}
    placeholder="Enter id"
    readOnly={false}
    type="text"
  />
  <br />
  <span
    className="bcgov-error"
    data-testid="email-error"
  />
</div>
`;

exports[`Input Component when mandatory and with label matches the snapshot (editable white) 1`] = `
<div>
  <label
    className="bcgov-label"
    htmlFor="textInputId"
  >
    <div
      className="bcgov-input-label"
    >
      Submission ID
    </div>
    <span
      className="bcgov-mandatory"
      id="asterisk"
    >
      *
    </span>
     
    <span
      className="bcgov-note"
    />
  </label>
  <input
    className="bcgov-editable-white"
    data-testid="input-test"
    id="textInputId"
    onChange={[Function]}
    placeholder="Enter id"
    readOnly={false}
    type="text"
  />
  <br />
  <span
    className="bcgov-error"
    data-testid="email-error"
  />
</div>
`;

exports[`Input Component when not mandatory and no label matches the snapshot (editable white) 1`] = `
<div>
  <input
    className="bcgov-editable-white"
    data-testid="input-test"
    id="textInputId"
    onChange={[Function]}
    placeholder="Enter id"
    readOnly={false}
    type="text"
  />
  <br />
  <span
    className="bcgov-error"
    data-testid="email-error"
  />
</div>
`;

exports[`Input Component when not mandatory and with label matches the snapshot (editable white) 1`] = `
<div>
  <label
    className="bcgov-label"
    htmlFor="textInputId"
  >
    <div
      className="bcgov-input-label"
    >
      Submission ID
    </div>
    
     
    <span
      className="bcgov-note"
    />
  </label>
  <input
    className="bcgov-editable-white"
    data-testid="input-test"
    id="textInputId"
    onChange={[Function]}
    placeholder="Enter id"
    readOnly={false}
    type="text"
  />
  <br />
  <span
    className="bcgov-error"
    data-testid="email-error"
  />
</div>
`;
